### PR TITLE
Add the ability to blacklist identifiers from a module

### PR DESCRIPTION
**Pull request by taktoa**
This allows you to prevent someone from importing a specific identifier from a module. This is useful if you want to, for example, maintain a policy that people never import identifiers that are reexported (e.g.: in MTL).

I am not sure how the hlint testing framework works, so I have not been able to test this code.

### CODE REVIEW

**Comment by googleson78** [comment]
You can run
`stack run hlint test`
to do tests (or the equivalent `cabal` command)

**Comment by ndmitchell** [comment]
Can you rebase/merge `master` into this so the tests will run properly.

**Comment by ndmitchell** [comment]
Can you give an example of the kind of thing this hint will spot, and not spot, so I can understand it a bit better? Does it only work with explicit imports?

**Comment by taktoa** [comment]
The point is that if you do `badidents: ["foo", "bar"]` then you won't be able to import `foo` and `bar` explicitly from that module. You can still import them implicitly, but in practice many codebases have a rule that disallows implicit imports.

**Comment by ndmitchell** [comment]
Did you try banning the functions with a hint with `Foo.bar` as the LHS, and "bad user" as the RHS? That way it should work even if they import qualified or don't use explicit import lists.

**Comment by taktoa** [comment]
I haven't tried that, but the goal here is to prevent users from importing reexported functions, so I am not sure if haskell-src-exts will resolve the reexport and ban the original identifier. In general, it just seems weird to hack the hint mechanism into providing a facility that users will naturally expect to be in the import management mechanism.

**Comment by taktoa** [comment]
Also, I think this technically allows you to ban explicit imports of type synonyms and pattern synonyms, and I don't think the hint system can do that.

**Comment by ndmitchell** [comment]
OK, that makes sense. Unfortunately there is nearly no testing for the restriction mechanism, so there's no good place to put it, so we can live dangerously and not have tests (which I've been fine with up until now - so can't complain about!) or as a follow up patch someone can write tests (which would be great, but you are under no obligation to do so).

It would be good to get just a sentence of two about this in the README though https://github.com/ndmitchell/hlint#restricting-items

File: src/Hint/Restrict.hs [hunk 1]
```diff
@@ -105,22 +109,52 @@ checkPragmas modu flags exts mps =
...
 
+importListToIdents :: IE GhcPs -> [String]
+importListToIdents =
+  catMaybes .
+  \case (IEVar _ n)              -> [fromName n]
+        (IEThingAbs _ n)         -> [fromName n]
+        (IEThingAll _ n)         -> [fromName n]
+        (IEThingWith _ n _ ns _) -> fromName n : map fromName ns
+        IEModuleContents {}      -> error "importListToIdents: this is only supposed to happen in an export list"
+        IEGroup {}               -> []
```
Comment (ndmitchell) [inline src/Hint/Restrict.hs]: Could we just use `_ -> []` for all these? I would rather be less tied to the GHC API if possible.

File: src/Hint/Restrict.hs [hunk 2]
```diff
@@ -105,22 +109,52 @@ checkPragmas modu flags exts mps =
...
     ]
 
+importListToIdents :: IE GhcPs -> [String]
+importListToIdents =
+  catMaybes .
+  \case (IEVar _ n)              -> [fromName n]
+        (IEThingAbs _ n)         -> [fromName n]
+        (IEThingAll _ n)         -> [fromName n]
+        (IEThingWith _ n _ ns _) -> fromName n : map fromName ns
+        IEModuleContents {}      -> error "importListToIdents: this is only supposed to happen in an export list"
```
Comment (ndmitchell) [inline src/Hint/Restrict.hs]: No need to error here.

**Comment by ndmitchell** [comment]
Thanks!

**Pull request merged**